ga4gh/backends.py

```diff
@@ -368,99 +311,104 @@ def getMetadata(self):
+            request, protocol.GASearchVariantSetsRequest,
+            protocol.GASearchVariantSetsResponse, "variantSets",
+            self.variantSetsGenerator)
+
+    def searchVariants(self, request):
+        """
+        Returns a GASearchVariantsResponse for the specified
+        GASearchVariantsRequest  object.
+        """
+        return self.runSearchRequest(
+            request, protocol.GASearchVariantsRequest,
+            protocol.GASearchVariantsResponse, "variants",
+            self.variantsGenerator)
+
+    def variantSetsGenerator(self, request):
+        """
+        Returns a generator over the (variantSet, nextPageToken) pairs defined
+        by the speficied request.
+        """
         currentIndex = 0
         if request.pageToken is not None:
             currentIndex = int(request.pageToken)
-        variantSetList = []
-        print(self._variantSetIds)
-        while (currentIndex < len(self._variantSetIds) and
-               len(variantSetList) < request.pageSize):
+        while currentIndex < len(self._variantSetIds):
             variantSet = protocol.GAVariantSet()
             variantSet.id = self._variantSetIds[currentIndex]
             variantSet.datasetId = "NotImplemented"
-            variantSet.metadata = self._variantSetMap[
+            variantSet.metadata = self._variantSetIdMap[
                 variantSet.id].getMetadata()
-            variantSetList.append(variantSet)
             currentIndex += 1
-        if currentIndex < len(self._variantSetIds):
-            response.nextPageToken = currentIndex
-        response.variantSets = variantSetList
-        return response
-
-
-class TabixBackend(Backend):
-    """
-    A class that serves variants indexed by tabix.
-    """
-    def __init__(self, dataDir):
-        Backend.__init__(self, dataDir, TabixVariantSet)
-
+            nextPageToken = None
+            if currentIndex < len(self._variantSetIds):
+                nextPageToken = str(currentIndex)
+            yield variantSet, nextPageToken
 
-class WormtableBackend(Backend):
-    """
-    A backend based on wormtable tables and indexes. This backend is provided
-    with a directory within which it can find VCF variant sets converted into
-    wormtable format. Each file within the specified directory is assumed to
-    be a wormtable, and the variant set ID will be the name of the
-    corresponding wormtable directory.
-    """
-    def __init__(self, dataDir):
-        Backend.__init__(self, dataDir, WormtableVariantSet)
+    def variantsGenerator(self, request):
+        """
+        Returns a generator over the (variant, nextPageToken) pairs defined by
+        the specified request.
+        """
+        variantSetIds = request.variantSetIds
+        startVariantSetIndex = 0
+        startPosition = request.start
+        if request.pageToken is not None:
+            startVariantSetIndex, startPosition = map(
+                int, request.pageToken.split(':'))
```

Comment:
This is a good idea. I've added a new method `parsePageToken` that verifies and parses the pageToken. It might work out that assuming all parts of the token must be integers is restrictive, but I think it's worth trying to do this for a start. It's a lot harder for an attacker to do nasty injection attacks when they are restricted to using integers.